cpp/cmake/EvalGpuArchs.cmake

```diff
@@ -0,0 +1,62 @@
+# Copyright (c) 2019, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+function(evaluate_gpu_archs gpu_archs)
+  set(eval_file ${PROJECT_BINARY_DIR}/eval_gpu_archs.cu)
+  set(eval_exe ${PROJECT_BINARY_DIR}/eval_gpu_archs)
+  set(error_file ${PROJECT_BINARY_DIR}/eval_gpu_archs.stderr.log)
+  file(WRITE ${eval_file}
+    "
+#include <cstdio>
+#include <set>
+#include <string>
+using namespace std;
+int main(int argc, char** argv) {
+  set<string> archs;
+  int nDevices;
+  if((cudaGetDeviceCount(&nDevices) == cudaSuccess) && (nDevices > 0)) {
+    for(int dev=0;dev<nDevices;++dev) {
+      char buff[32];
+      cudaDeviceProp prop;
+      if(cudaGetDeviceProperties(&prop, dev) != cudaSuccess) continue;
+      sprintf(buff, \"%d%d\", prop.major, prop.minor);
+      archs.insert(buff);
+    }
+  }
+  if(archs.empty()) {
+    printf(\"ALL\");
+  } else {
+    bool first = true;
+    for(const auto& arch : archs) {
+      printf(first? \"%s\" : \";%s\", arch.c_str());
+      first = false;
+    }
+  }
+  printf(\"\\n\");
+  return 0;
+}
+")
```

Comment:
I really, really, don't like this. Can we check in a Python file or something else instead so we don't have to compile to detect options to compile?

Reply:
Python script I was dabbling with:
```python
from ctypes import *

libcuda = CDLL("libcuda.so")

def get_null_char_pos(string_stream):
    i = 0
    while True:
        if string_stream[i] == b'\x00':
            return i
        i += 1

def check_error(result):
    if result == 0:
        return None
    else:
        error_name = POINTER(POINTER(c_char))(c_char(0))
        error_result = libcuda.cuGetErrorName(result, error_name)
        check_error(error_result)
        length = get_null_char_pos(error_name.contents)
        error_name = error_name.contents[0:length]

        error_string = POINTER(POINTER(c_char))(c_char(0))
        error_result = libcuda.cuGetErrorString(result, error_string)
        check_error(error_result)
        length = get_null_char_pos(error_string.contents)
        error_string = error_string.contents[0:length]

        raise RuntimeError("[{0}] {1}: {2}".format(str(result), error_name.decode(), error_string.decode()))

result = libcuda.cuInit(0)
check_error(result)

major = POINTER(c_int)(c_int())
minor = POINTER(c_int)(c_int())

result = libcuda.cuDeviceComputeCapability(major, minor, 0)
check_error(result)

print(str(major.contents.value)+str(minor.contents.value))
```

Reply:
I'm fine either way :man_shrugging:

Reply:
I find the C script simpler and easier to understand for CUDA developers. This Python script doesn't seem to do the same thing -- it doesn't iterate over all devices.

Reply:
Seems like this is something that could be provided by cmake?
See https://gitlab.kitware.com/cmake/cmake/issues/19199 and the latest comment...

Reply:
Seems like this is something that could be upstreamed to cmake?  
See also https://gitlab.kitware.com/cmake/cmake/issues/19199...